<a href="https://colab.research.google.com/github/Prudycruise/nhs-healthcare-analytics/blob/main/python/01_data_inspection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd

url = "https://raw.githubusercontent.com/Prudycruise/nhs-healthcare-analytics/main/data/raw/AE_attendances_england_monthly.csv"

df = pd.read_csv(url)

print("Dataset loaded successfully")
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

df.head()

Dataset loaded successfully
Rows: 27112
Columns: 20


,Unnamed: 0,date,Name,Type 1 Departments - Major A&E,Type 2 Departments - Single Specialty,Type 3 Departments - Other A&E/Minor Injury Unit,Total attendances,Type 1 Departments - 4 hours to decision,Type 2 Departments - 4 hours to decision,Type 3 Departments - 4 hours to decision,Percentage in 4 hours or less (all),Emergency Admissions via Type 1 A&E in 4 hours,Emergency Admissions via Type 2 A&E in 4 hours,Emergency Admissions via Type 3 and 4 A&E in 4 hours,Other Emergency admissions (i.e not via A&E),Number of patients spending >12 hours from decision to admit to admission,month,year,lat,lon
0,0,2010-11-01,Aintree University Hospitals NHS Foundation Trust,4622.0,0.0,0.0,4622.0,7.0,0.0,0.0,2.995435,1406.0,0.0,0.0,530.0,0.0,44.0,8040.0,53.461606,-2.943427
1,1,2010-11-01,Airedale NHS Trust,3965.0,0.0,0.0,3965.0,65.0,0.0,0.0,3.934107,892.0,0.0,0.0,568.0,0.0,44.0,8040.0,NaN,NaN
2,2,2010-11-01,Alder Hey Children’S NHS Foundation Trust,4541.0,0.0,0.0,4541.0,71.0,0.0,0.0,3.937936,1825.0,0.0,0.0,270.0,0.0,44.0,8040.0,50.183000,-5.416000
3,3,2010-11-01,Ashford And St Peter'S Hospitals NHS Trust,7010.0,0.0,769.0,7779.0,392.0,0.0,0.0,3.797332,1902.0,0.0,0.0,163.0,0.0,44.0,8040.0,51.655000,-0.395694
4,4,2010-11-01,"Ashton, Leigh And Wigan Primary Care Trust",0.0,0.0,4297.0,4297.0,0.0,0.0,0.0,4.000000,0.0,0.0,0.0,0.0,0.0,44.0,8040.0,51.272000,0.529000


In [3]:
print("Column names:")

for column in df.columns:
    print(column)

Column names:
Unnamed: 0
date
Name
Type 1 Departments - Major A&E
Type 2 Departments - Single Specialty
Type 3 Departments - Other A&E/Minor Injury Unit
Total attendances
Type 1 Departments - 4 hours to decision
Type 2 Departments - 4 hours to decision
Type 3 Departments - 4 hours to decision
Percentage in 4 hours or less (all)
Emergency Admissions via Type 1 A&E in 4 hours
Emergency Admissions via Type 2 A&E in 4 hours
Emergency Admissions via Type 3 and 4 A&E in 4 hours
Other Emergency admissions (i.e not via A&E)
Number of patients spending >12 hours from decision to admit to admission
month
year
lat
lon


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27112 entries, 0 to 27111
Data columns (total 20 columns):
 #   Column                                                                     Non-Null Count  Dtype  
---  ------                                                                     --------------  -----  
 0   Unnamed: 0                                                                 27112 non-null  int64  
 1   date                                                                       27112 non-null  object 
 2   Name                                                                       27108 non-null  object 
 3   Type 1 Departments - Major A&E                                             27112 non-null  float64
 4   Type 2 Departments - Single Specialty                                      27112 non-null  float64
 5   Type 3 Departments - Other A&E/Minor Injury Unit                           27112 non-null  float64
 6   Total attendances                                     

In [5]:
missing_values = df.isnull().sum()

print("Missing values:")
print(missing_values[missing_values > 0])

Missing values:
Name       4
lat     8010
lon     8010
dtype: int64


In [6]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [7]:
df["date"] = pd.to_datetime(df["date"], errors="coerce")

print("Earliest date:", df["date"].min())
print("Latest date:", df["date"].max())
print("Invalid dates:", df["date"].isnull().sum())

Earliest date: 2010-11-01 00:00:00
Latest date: 2020-06-01 00:00:00
Invalid dates: 0


In [8]:
incorrect_months = df[df["month"] != df["date"].dt.month]
incorrect_years = df[df["year"] != df["date"].dt.year]

print("Rows with incorrect months:", len(incorrect_months))
print("Rows with incorrect years:", len(incorrect_years))

Rows with incorrect months: 12320
Rows with incorrect years: 12320


In [9]:
percentage_column = "Percentage in 4 hours or less (all)"

print(df[percentage_column].describe())
print()
print("Values above 100%:", (df[percentage_column] > 1).sum())

count    27112.000000
mean         2.234394
std          1.665038
min          0.000000
25%          0.910265
50%          1.000000
75%          3.941170
max          5.000000
Name: Percentage in 4 hours or less (all), dtype: float64

Values above 100%: 11792


## Initial Data Quality Findings

- The dataset contains 27,112 rows and 20 columns.
- The date range runs from November 2010 to June 2020.
- The `Unnamed: 0` column appears to be an unnecessary index.
- Some organisation names are missing.
- Many latitude and longitude values are missing.
- Some month and year values do not match the date.
- Some four-hour performance values exceed 100%.
- Suspicious records require investigation before analysis.
- The raw dataset will remain unchanged.

## Analysis Questions

1. How did total A&E attendance change over time?
2. Which NHS organisations recorded the highest total attendance?
3. How did attendance differ between Type 1, Type 2 and Type 3 departments?
4. How did four-hour performance change over time?
5. Which organisations recorded the highest numbers of patients waiting longer than 12 hours?